### DISCUSSION TITLE
Сводная таблица

### DISCUSSION THREAD

**Discussion by juramebel** [totumonline/totum-ru-issues-and-discussions#68] [Как это сделать?] [answered]
Здравствуйте. 
Для отчетов и разных временных таблиц постоянно нужен функционал типа "Сводная таблица".
Как проще его реализовать.
Пока приходят на ум два пути.
1. Собираем данные в rowlist, потом какими то сложными манипуляциями вытаскиваем уникальные значения, через while создаем какой то новый уже rowlist, который используем в поле insert
2. Также вытаскиваем уникальные данные, вставляем с ними строчки, остальные поля заполняются через selectlist из таблиц источников...
Но в общем и то и другое выглядит заморочено, достаточно много кода.
Может быть есть какой то простой путь, который я не вижу?

**Comment by ttm-alexeypolunin** [comment, marked as answer]
Ответ далее в обсуждении к этому посту

```Не пропустите кнопочку [view more], там есть полезное!```

**Reply by juramebel** [reply]
Ну.. я может слишком широко даже хватанул.
Ок, не сводная таблица (хотя ее в классическом понимании какого нибудь экселя, тоже хотелось бы)
Допустим сейчас у меня такая задача.
Есть простая таблица с заявками вида:
позиция / кол-во / связанный объект

Мне нужно агрегировать данные из этих записей. 
Ну то есть вместо разных записей с одной и той же позицией, сделать одну и просуммировать все колва по ней.

**Reply by juramebel** [reply]
![image](https://user-images.githubusercontent.com/19141501/152112010-3190828a-b5f0-4e37-8104-b69dcc1f6576.png)

**Reply by ttm-alexeypolunin** [reply]
Вот это делается через автозаполнение по `insert`. Те мы берем список по `позиции`, делаем его уникальным и суем в `insert`:

```
=: listUniq(list: $list)
list: selectList(table: 'table'; field: 'poz')

```

И потом в Сумма по полю делаем код с суммированием:

```
=: listSum(list: $list)
list: selectList(table: 'table'; field: 'kol'; where: 'poz' = #insert)
```

**Reply by juramebel** [reply]
Это да... Это так скажем, максимально упрощенная ситуация. 
А вот скажем, если есть третья колонка.
В ней дата поставки требуемой. 
Как сделать уже настоящую сводную таблицу 

![image](https://user-images.githubusercontent.com/19141501/152122343-ba626ce3-181c-4886-9fbc-2f833cac5046.png)

**Reply by ttm-alexeypolunin** [reply]
Все тоже-самое, только вам нужно 3 колонки создать и в добавить дополнительный `where` в выборку для сумм:

```
=: listSum(list: $list)
list: selectList(table: 'table'; field: 'kol'; where: 'poz' = #insert; where: 'week' = 1)
```

Чего Тотум не умеет, но возможно будет уметь в `Pro` — это динамические колонки. Те сейчас мы обсуждаем, что колонки в этом отчете добавляются руками, но может быть такое, что их разное количество в зависимости от условий и в продвинутой версии такой функционал запланирован.

Но он все равно не будет повторять Excel.

**Reply by juramebel** [reply]
Все. Теперь все предельно ясно.
Тогда последний вопрос.
А как это сделать не из таблицы, а имея, к примеру, вытащенные в rowlist данные из циклов? 
То есть у нас на руках списки списков, можно ли их сразу преобразовать в такую таблицу, без промежуточной.
Подозреваю, что да, но именно по первому варианту, который я описывал. 
Который для меня в плане кода становится каким то запутанным...

**Reply by ttm-alexeypolunin** [reply]
Я наверное видос запишу и сделаю более подробную инструкцию

**Reply by ttm-alexeypolunin** [reply]
### [Видео на Youtube —>](https://youtu.be/BDcd4a4AJ2c)
